CUDA compilation failing with VS 2022 Community

### Issue description

CUDA compilation fails with code

### Expected Behavior

Compilation with CUDA should work

### Actual Behavior

I have an error MSB8020 whan trying to recompile with CUDA.
I tested to compile with CUDA direclty on the llama.ccp project with CMAKE, without error.

### Steps to reproduce

Lauching `npx --no node-llama-cpp download --cuda`

### My Environment

| Dependency               | Version             |
| ---                      | ---                 |
| Operating System         |          Windows 11           |
| CPU                      | AMD Ryzen 5 3600 6-Core Processor |
| Node.js version          | v18.18.2             |
| `node-llama-cpp` version | 2.7.3             |
| CUDA version | v12.2            |
| Visual Studio version | 2022 community           |

### Additional Context

The logs : 

```
vuith@THOMZ-FIXE MINGW64 ~/GitRepositories/node-llama-cpp-experiment (main)
$ npx --no node-llama-cpp download --cuda
Repo: ggerganov/llama.cpp
Release: b1378
CUDA: enabled

✔ Fetched llama.cpp info
✔ Removed existing llama.cpp directory
Cloning llama.cpp
Clone ggerganov/llama.cpp (local bundle)  100% ████████████████████████████████████████  0s
◷ Compiling llama.cpp
Not searching for unused variables given on the command line.
CMake Error at CMakeLists.txt:3 (project):
  Failed to run MSBuild command:

    C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe

  to get the value of VCTargetsPath:

    Version MSBuild 17.7.2+d6990bcfa pour .NET Framework
    La génération a démarré 22/10/2023 18:51:39.

    Projet "C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\CMakeFiles\3.28.0-rc2\VCTargetsPath.vcxproj" sur le noud 1 (cibles par défaut).
    C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(456,5): error MSB8020: Les outils de génération pour C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2 (ensemble d'outils de plateforme = 'C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2') sont introuvables. Pour générer à l'aide des outils de génération C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2, installez les outils de génération C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2. Vous avez également la possibilité de mettre à niveau les outils Visual Studio actuels en sélectionnant le menu Projet ou en cliquant avec le bouton droit sur la solution, puis en sélectionnant "Recibler la solution". [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\CMakeFiles\3.28.0-rc2\VCTargetsPath.vcxproj]
    Génération du projet "C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\CMakeFiles\3.28.0-rc2\VCTargetsPath.vcxproj" terminée (cibles par défaut) -- ÉCHEC
```

### Relevant Features Used

- [ ] Metal support
- [X] CUDA support
- [ ] Grammar

### Are you willing to resolve this issue by submitting a Pull Request?

No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

Yes, this one's working fine : 

```
vuith@THOMZ-FIXE MINGW64 ~/GitRepositories/node-llama-cpp-experiment (main)
$ npx --no node-llama-cpp download
Repo: ggerganov/llama.cpp
Release: b1378

✔ Fetched llama.cpp info
✔ Removed existing llama.cpp directory
Cloning llama.cpp
Clone ggerganov/llama.cpp (local bundle)  100% ████████████████████████████████████████  0s
◷ Compiling llama.cpp
Not searching for unused variables given on the command line.
-- The C compiler identification is MSVC 19.37.32825.0
-- The CXX compiler identification is MSVC 19.37.32825.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.42.0.windows.2")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM: x64
-- x86 detected
-- Configuring done (4.8s)
-- Generating done (0.1s)
-- Build files have been written to: C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/build
Version MSBuild 17.7.2+d6990bcfa pour .NET Framework

  1>Checking Build System
  Generating build details from Git
  -- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.42.0.windows.2")
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  ggml.c
  ggml-alloc.c
  ggml-backend.c
  k_quants.c
  Génération de code en cours...
  ggml.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llam
  a.cpp\ggml.dir\Release\ggml.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  llama.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1054,31)
: warning C4305: 'initialisation' : troncation de 'double' à 'float' [C:\Users\vuith\GitRepositories\node-llama-cpp-exp
eriment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1202,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u0120' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1203,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1204,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u0120' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1205,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(2235,60)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(9452,28)
: warning C4146: opérateur moins unaire appliqué à un type non signé, le résultat sera non signé [C:\Users\vuith\GitRep
ositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(9482,28)
: warning C4146: opérateur moins unaire appliqué à un type non signé, le résultat sera non signé [C:\Users\vuith\GitRep
ositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
  llama.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\lla
  ma.cpp\Release\llama.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/common/CMakeLists.txt
  common.cpp
  sampling.cpp
  console.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\console.cpp
(253,30): warning C4267: 'initialisation' : conversion de 'size_t' en 'DWORD', perte possible de données [C:\Users\vuit
h\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\console.cpp
(407,28): warning C4267: 'initialisation' : conversion de 'size_t' en 'int', perte possible de données [C:\Users\vuith\
GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
  grammar-parser.cpp
  train.cpp
  Génération de code en cours...
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\common.cpp(
785): warning C4715: 'gpt_random_prompt' : les chemins de contrôle ne retournent pas tous une valeur [C:\Users\vuith\Gi
tRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
  common.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\ll
  ama.cpp\common\common.dir\Release\common.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  ggml_static.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\bui
  ld\llama.cpp\Release\ggml_static.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/CMake
  Lists.txt
  addon.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\log.h(379,1
9): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable depreca
tion, use _CRT_SECURE_NO_WARNINGS. See online help for details. [C:\Users\vuith\GitRepositories\node-llama-cpp-experime
nt\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\log.h(387,9
7): warning C4996: 'strerror': This function or variable may be unsafe. Consider using strerror_s instead. To disable d
eprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. [C:\Users\vuith\GitRepositories\node-llama-cpp-ex
periment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(339,53): warning C
4267: 'argument' : conversion de 'size_t' en 'int32_t', perte possible de données [C:\Users\vuith\GitRepositories\node-
llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(341,33): warning C
4267: '=' : conversion de 'size_t' en 'int32_t', perte possible de données [C:\Users\vuith\GitRepositories\node-llama-c
pp-experiment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(415,9): warning C4
996: 'llama_sample_temperature': use llama_sample_temp instead [C:\Users\vuith\GitRepositories\node-llama-cpp-experimen
t\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
  win_delay_load_hook.cc
  Génération de code en cours...
     Création de la bibliothèque C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/l
  lama/build/Release/llama-addon.lib et de l'objet C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_module
  s/node-llama-cpp/llama/build/Release/llama-addon.exp
  llama-addon.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\bui
  ld\Release\llama-addon.node
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/CMake
  Lists.txt
✔ Compiled llama.cpp


Repo: ggerganov/llama.cpp
Release: b1378

Done
```

Thanks for the tip @sk-1982 ! I just tested to remove this line of code in `dist\utils\compileLLamaCpp.js` (in my node_modules) and it worked : 


```
vuith@THOMZ-FIXE MINGW64 ~/GitRepositories/node-llama-cpp-experiment (main)
$ npx --no node-llama-cpp download --cuda
Repo: ggerganov/llama.cpp
Release: b1378
CUDA: enabled

✔ Fetched llama.cpp info
✔ Removed existing llama.cpp directory
Cloning llama.cpp
Clone ggerganov/llama.cpp (local bundle)  100% ████████████████████████████████████████  0s
◷ Compiling llama.cpp
Not searching for unused variables given on the command line.
-- The C compiler identification is MSVC 19.37.32825.0
-- The CXX compiler identification is MSVC 19.37.32825.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.42.0.windows.2")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Found CUDAToolkit: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.2/include (found version "12.2.140")
-- cuBLAS found
-- The CUDA compiler identification is NVIDIA 12.2.140
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.2/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Using CUDA architectures: 52;61;70
-- CMAKE_SYSTEM_PROCESSOR: AMD64
-- CMAKE_GENERATOR_PLATFORM: x64
-- x86 detected
-- Configuring done (22.1s)
-- Generating done (0.1s)
-- Build files have been written to: C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/build
Version MSBuild 17.7.2+d6990bcfa pour .NET Framework

  1>Checking Build System
  Generating build details from Git
  -- Found Git: C:/Program Files/Git/mingw64/bin/git.exe (found version "2.42.0.windows.2")
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  Compiling CUDA source file ..\..\llama.cpp\ggml-cuda.cu...

  C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp>"C:\Progra
  m Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2\bin\nvcc.exe"  --use-local-env -ccbin "C:\Program Files\Microsoft Vis
  ual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64" -x cu   -I"C:\Users\vuith\GitRepositories\node-l
  lama-cpp-experiment\node_modules\node-addon-api" -I"C:\Users\vuith\.cmake-js\node-x64\v18.18.2\include\node" -I"C:\Us
  ers\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\." -I"C:\Program File
  s\NVIDIA GPU Computing Toolkit\CUDA\v12.2\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2\includ
  e"     --keep-dir x64\Release -use_fast_math -maxrregcount=0   --machine 64 --compile -cudart static --generate-code=
  arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=
  compute_70,code=[compute_70,sm_70] -Xcompiler="/EHsc -Ob2"   -D_WINDOWS -DNDEBUG -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS
   -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SI
  ZE=128 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"" -D_MBCS -DWIN32 -D_WINDOWS -DNDEBUG
   -DNAPI_VERSION=7 -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITER
  ATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128 -D_CRT_SECURE_NO_WARNINGS -D_XOPEN_SOURCE=600 -D"CMAKE_INTDIR=\"Release\"
  " -Xcompiler "/EHsc /W3 /nologo /O2 /FS   /MD /GR" -Xcompiler "/Fdggml.dir\Release\ggml.pdb" -o ggml.dir\Release\ggml
  -cuda.obj "C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\ggml-
  cuda.cu"
  ggml-cuda.cu
  tmpxft_00004eb8_00000000-7_ggml-cuda.compute_70.cudafe1.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu(5668)
: warning C4477: 'fprintf' : la chaîne de format '%ld' nécessite un argument de type 'long', mais l'argument variadique
 1 est de type 'int64_t' [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\bu
ild\llama.cpp\ggml.vcxproj]
  C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu(566
  8): note: utilisez '%lld' dans la chaîne de format
  C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu(566
  8): note: utilisez '%Id' dans la chaîne de format
  C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\ggml-cuda.cu(566
  8): note: utilisez '%I64d' dans la chaîne de format
  ggml.c
  ggml-alloc.c
  ggml-backend.c
  k_quants.c
  Génération de code en cours...
  ggml.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llam
  a.cpp\ggml.dir\Release\ggml.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  llama.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1054,31)
: warning C4305: 'initialisation' : troncation de 'double' à 'float' [C:\Users\vuith\GitRepositories\node-llama-cpp-exp
eriment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1202,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u0120' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1203,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1204,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u0120' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(1205,40)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(2235,60)
: warning C4566: le caractère représenté par le nom de caractère universel '\u010A' ne peut pas être représenté dans la
 page de codes actuelle (1252) [C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\ll
ama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(9452,28)
: warning C4146: opérateur moins unaire appliqué à un type non signé, le résultat sera non signé [C:\Users\vuith\GitRep
ositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\llama.cpp(9482,28)
: warning C4146: opérateur moins unaire appliqué à un type non signé, le résultat sera non signé [C:\Users\vuith\GitRep
ositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\llama.vcxproj]
  llama.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\lla
  ma.cpp\Release\llama.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/common/CMakeLists.txt
  common.cpp
  sampling.cpp
  console.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\console.cpp
(253,30): warning C4267: 'initialisation' : conversion de 'size_t' en 'DWORD', perte possible de données [C:\Users\vuit
h\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\console.cpp
(407,28): warning C4267: 'initialisation' : conversion de 'size_t' en 'int', perte possible de données [C:\Users\vuith\
GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
  grammar-parser.cpp
  train.cpp
  Génération de code en cours...
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\common.cpp(
785): warning C4715: 'gpt_random_prompt' : les chemins de contrôle ne retournent pas tous une valeur [C:\Users\vuith\Gi
tRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama.cpp\common\common.vcxproj]
  common.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\ll
  ama.cpp\common\common.dir\Release\common.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/llama
  .cpp/CMakeLists.txt
  ggml_static.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\bui
  ld\llama.cpp\Release\ggml_static.lib
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/CMake
  Lists.txt
  addon.cpp
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\log.h(379,1
9): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable depreca
tion, use _CRT_SECURE_NO_WARNINGS. See online help for details. [C:\Users\vuith\GitRepositories\node-llama-cpp-experime
nt\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\llama.cpp\common\log.h(387,9
7): warning C4996: 'strerror': This function or variable may be unsafe. Consider using strerror_s instead. To disable d
eprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. [C:\Users\vuith\GitRepositories\node-llama-cpp-ex
periment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(339,53): warning C
4267: 'argument' : conversion de 'size_t' en 'int32_t', perte possible de données [C:\Users\vuith\GitRepositories\node-
llama-cpp-experiment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(341,33): warning C
4267: '=' : conversion de 'size_t' en 'int32_t', perte possible de données [C:\Users\vuith\GitRepositories\node-llama-c
pp-experiment\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\addon.cpp(415,9): warning C4
996: 'llama_sample_temperature': use llama_sample_temp instead [C:\Users\vuith\GitRepositories\node-llama-cpp-experimen
t\node_modules\node-llama-cpp\llama\build\llama-addon.vcxproj]
  win_delay_load_hook.cc
  Génération de code en cours...
     Création de la bibliothèque C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/l
  lama/build/Release/llama-addon.lib et de l'objet C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_module
  s/node-llama-cpp/llama/build/Release/llama-addon.exp
  llama-addon.vcxproj -> C:\Users\vuith\GitRepositories\node-llama-cpp-experiment\node_modules\node-llama-cpp\llama\bui
  ld\Release\llama-addon.node
  Building Custom Rule C:/Users/vuith/GitRepositories/node-llama-cpp-experiment/node_modules/node-llama-cpp/llama/CMake
  Lists.txt
✔ Compiled llama.cpp


Repo: ggerganov/llama.cpp
Release: b1378

Done
```

--- a/.commitlintrc.json
+++ b/.commitlintrc.json
@@ -1,3 +1,6 @@
 {
-    "extends": ["@commitlint/config-conventional"]
+    "extends": ["@commitlint/config-conventional"],
+    "rules": {
+        "subject-case": [0, "never"]
+    }
 }

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -231,6 +231,9 @@ jobs:
       - name: Install modules
         run: npm ci
 
+      - name: Build binary
+        run: node ./dist/cli/cli.js build
+
       - name: Run standalone tests
         run: npm run test:standalone
 

--- a/docs/guide/CUDA.md
+++ b/docs/guide/CUDA.md
@@ -23,20 +23,38 @@ To build `node-llama-cpp` with any of these options, set an environment variable
 ### Fix the `Failed to detect a default CUDA architecture` build error
 To fix this issue you have to set the `CUDACXX` environment variable to the path of the `nvcc` compiler.
 
-For example, if you installed CUDA Toolkit 12.2 on Windows, you have to run the following command:
-```bash
-set CUDACXX=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2\bin\nvcc.exe
-```
-
-On Linux, it would be something like this:
-```bash
+For example, if you have installed CUDA Toolkit 12.2, you have to run a command like this:
+::: code-group
+```bash [Linux]
 export CUDACXX=/usr/local/cuda-12.2/bin/nvcc
 ```
+
+```bash [Windows]
+set CUDACXX=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.2\bin\nvcc.exe
+```
+:::
 
 Then run the build command again to check whether setting the `CUDACXX` environment variable fixed the issue.
 
 ### Fix the `The CUDA compiler identification is unknown` build error
 The solution to this error is the same as [the solution to the `Failed to detect a default CUDA architecture` error](#fix-the-failed-to-detect-a-default-cuda-architecture-build-error).
+
+### Fix the `A single input file is required for a non-link phase when an outputfile is specified` build error
+To fix this issue you have to set the `CMAKE_GENERATOR_TOOLSET` cmake option to the CUDA home directory, usually already set as the `CUDA_PATH` environment variable.
+
+To do this, set the `NODE_LLAMA_CPP_CMAKE_OPTION_CMAKE_GENERATOR_TOOLSET` environment variable to the path of your CUDA home directory:
+
+::: code-group
+```bash [Linux]
+export NODE_LLAMA_CPP_CMAKE_OPTION_CMAKE_GENERATOR_TOOLSET=$CUDA_PATH
+```
+
+```bash [Windows]
+set NODE_LLAMA_CPP_CMAKE_OPTION_CMAKE_GENERATOR_TOOLSET=%CUDA_PATH%
+```
+:::
+
+Then run the build command again to check whether setting the `CMAKE_GENERATOR_TOOLSET` cmake option fixed the issue.
 
 ## Using `node-llama-cpp` with CUDA
 After you build `node-llama-cpp` with CUDA support, you can use it normally.

--- a/docs/guide/CUDA.md
+++ b/docs/guide/CUDA.md
@@ -35,6 +35,9 @@ export CUDACXX=/usr/local/cuda-12.2/bin/nvcc
 
 Then run the build command again to check whether setting the `CUDACXX` environment variable fixed the issue.
 
+### Fix the `The CUDA compiler identification is unknown` build error
+The solution to this error is the same as [the solution to the `Failed to detect a default CUDA architecture` error](#fix-the-failed-to-detect-a-default-cuda-architecture-build-error).
+
 ## Using `node-llama-cpp` with CUDA
 After you build `node-llama-cpp` with CUDA support, you can use it normally.
 

--- a/docs/guide/building-from-source.md
+++ b/docs/guide/building-from-source.md
@@ -27,6 +27,8 @@ If `cmake` is not installed on your machine, `node-llama-cpp` will automatically
 
 If the build fails, make sure you have the required dependencies of `cmake` installed on your machine. More info is available [here](https://github.com/cmake-js/cmake-js#:~:text=projectRoot/build%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Bstring%5D-,Requirements%3A,-CMake) (you don't have to install `cmake` or `cmake-js`, just the dependencies).
 
+If the build fails on macOS with the error `"/usr/bin/cc" is not able to compile a simple test program`, try running `xcode-select --install` to install the Xcode command line tools.
+
 :::
 
 ## `download` and `build` commands

--- a/docs/guide/cli/build.md
+++ b/docs/guide/cli/build.md
@@ -10,6 +10,10 @@ const commandDoc = docs.build;
 
 {{commandDoc.description}}
 
+::: info
+If the build fails on macOS with the error `"/usr/bin/cc" is not able to compile a simple test program`, try running `xcode-select --install` to install the Xcode command line tools.
+:::
+
 ## Usage
 ```shell-vue
 {{commandDoc.usage}}

--- a/docs/guide/cli/download.md
+++ b/docs/guide/cli/download.md
@@ -20,6 +20,10 @@ This is useful for building from source on machines that aren't connected to the
 
 :::
 
+::: info
+If the build fails on macOS with the error `"/usr/bin/cc" is not able to compile a simple test program`, try running `xcode-select --install` to install the Xcode command line tools.
+:::
+
 ## Usage
 ```shell-vue
 {{commandDoc.usage}}

--- a/docs/guide/index.md
+++ b/docs/guide/index.md
@@ -14,6 +14,11 @@ npm install --save node-llama-cpp
 > If binaries are not available for your platform, it'll fallback to download a release of `llama.cpp` and build it from source with `cmake`.
 > To disable this behavior, set the environment variable `NODE_LLAMA_CPP_SKIP_DOWNLOAD` to `true`.
 
+## ESM usage
+`node-llama-cpp` is an [ES module](https://nodejs.org/api/esm.html#modules-ecmascript-modules), so can only use `import` to load it and cannot use `require`.
+
+To make sure you can use it in your project, make sure your `package.json` file has `"type": "module"` in it.
+
 ## CUDA and Metal support
 **Metal:** Metal support is enabled by default on macOS. If you're using a Mac with an Intel chip, [you might want to disable it](./Metal.md).
 

--- a/llama/CMakeLists.txt
+++ b/llama/CMakeLists.txt
@@ -1,6 +1,6 @@
-cmake_minimum_required(VERSION 3.12)
+cmake_minimum_required(VERSION 3.13)
 
-project ("llama-addon")
+project("llama-addon" C CXX)
 
 if (MSVC)
   # add_compile_options(/EHsc)

--- a/llama/addon.cpp
+++ b/llama/addon.cpp
@@ -208,13 +208,13 @@ class LLAMAContext : public Napi::ObjectWrap<LLAMAContext> {
     return Napi::String::New(info.Env(), ss.str());
   }
   Napi::Value TokenBos(const Napi::CallbackInfo& info) {
-    return Napi::Number::From(info.Env(), llama_token_bos(ctx));
+    return Napi::Number::From(info.Env(), llama_token_bos(model->model)); // TODO: move this to the model
   }
   Napi::Value TokenEos(const Napi::CallbackInfo& info) {
-    return Napi::Number::From(info.Env(), llama_token_eos(ctx));
+    return Napi::Number::From(info.Env(), llama_token_eos(model->model)); // TODO: move this to the model
   }
   Napi::Value TokenNl(const Napi::CallbackInfo& info) {
-    return Napi::Number::From(info.Env(), llama_token_nl(ctx));
+    return Napi::Number::From(info.Env(), llama_token_nl(model->model)); // TODO: move this to the model
   }
   Napi::Value GetContextSize(const Napi::CallbackInfo& info) {
     return Napi::Number::From(info.Env(), llama_n_ctx(ctx));
@@ -223,7 +223,7 @@ class LLAMAContext : public Napi::ObjectWrap<LLAMAContext> {
     int token = info[0].As<Napi::Number>().Int32Value();
     std::stringstream ss;
 
-    const char* str = llama_token_get_text(ctx, token);
+    const char* str = llama_token_get_text(model->model, token); // TODO: move this to the model
     if (str == nullptr) {
       return info.Env().Undefined();
     }
@@ -377,7 +377,7 @@ class LLAMAContextEvalWorker : Napi::AsyncWorker, Napi::Promise::Deferred {
 
     llama_token_data_array candidates_p = { candidates.data(), candidates.size(), false };
 
-    auto eos_token = llama_token_eos(ctx->ctx);
+    auto eos_token = llama_token_eos(ctx->model->model);
 
     if (use_repeat_penalty && !repeat_penalty_tokens.empty()) {
       llama_sample_repetition_penalties(

--- a/llama/addon.cpp
+++ b/llama/addon.cpp
@@ -336,18 +336,14 @@ class LLAMAContextEvalWorker : Napi::AsyncWorker, Napi::Promise::Deferred {
 
   protected:
   void Execute() {
-    llama_batch batch = llama_batch_init(tokens.size(), 0);
+    llama_batch batch = llama_batch_init(tokens.size(), 0, 1);
 
-    batch.n_tokens = tokens.size();
-
-    for (int32_t i = 0; i < batch.n_tokens; i++) {
-        batch.token[i]  = tokens[i];
-        batch.pos[i]    = ctx->n_cur;
-        batch.seq_id[i] = 0;
-        batch.logits[i] = false;
+    for (size_t i = 0; i < tokens.size(); i++) {
+        llama_batch_add(batch, tokens[i], ctx->n_cur, { 0 }, false);
 
         ctx->n_cur++;
     }
+    GGML_ASSERT(batch.n_tokens == (int) tokens.size());
 
     batch.logits[batch.n_tokens - 1] = true;
 
@@ -384,11 +380,8 @@ class LLAMAContextEvalWorker : Napi::AsyncWorker, Napi::Promise::Deferred {
     auto eos_token = llama_token_eos(ctx->ctx);
 
     if (use_repeat_penalty && !repeat_penalty_tokens.empty()) {
-      llama_sample_repetition_penalty(
-        ctx->ctx, &candidates_p, repeat_penalty_tokens.data(), repeat_penalty_tokens.size(), repeat_penalty
-      );
-      llama_sample_frequency_and_presence_penalties(
-        ctx->ctx, &candidates_p, repeat_penalty_tokens.data(), repeat_penalty_tokens.size(),
+      llama_sample_repetition_penalties(
+        ctx->ctx, &candidates_p, repeat_penalty_tokens.data(), repeat_penalty_tokens.size(), repeat_penalty,
         repeat_penalty_frequency_penalty, repeat_penalty_presence_penalty
       );
     }

--- a/llama/package.json
+++ b/llama/package.json
@@ -1,0 +1,5 @@
+{
+  "binary": {
+    "napi_versions": [7]
+  }
+}

--- a/package.json
+++ b/package.json
@@ -100,7 +100,6 @@
     "@commitlint/cli": "^17.7.1",
     "@commitlint/config-conventional": "^17.7.0",
     "@semantic-release/exec": "^6.0.3",
-    "@types/bytes": "^3.1.1",
     "@types/cli-progress": "^3.11.0",
     "@types/cross-spawn": "^6.0.2",
     "@types/fs-extra": "^11.0.1",

--- a/src/cli/commands/DownloadCommand.ts
+++ b/src/cli/commands/DownloadCommand.ts
@@ -13,7 +13,10 @@ import {setBinariesGithubRelease} from "../../utils/binariesGithubRelease.js";
 import {downloadCmakeIfNeeded} from "../../utils/cmake.js";
 import withStatusLogs from "../../utils/withStatusLogs.js";
 import {getIsInDocumentationMode} from "../../state.js";
-import {unshallowAndSquashCurrentRepoAndSaveItAsReleaseBundle} from "../../utils/gitReleaseBundles.js";
+import {
+    getGitBundlePathForRelease,
+    unshallowAndSquashCurrentRepoAndSaveItAsReleaseBundle
+} from "../../utils/gitReleaseBundles.js";
 import {cloneLlamaCppRepo} from "../../utils/cloneLlamaCppRepo.js";
 
 type DownloadCommandArgs = {
@@ -91,6 +94,7 @@ export const DownloadCommand: CommandModule<object, DownloadCommandArgs> = {
 export async function DownloadLlamaCppCommand({
     repo, release, arch, nodeTarget, metal, cuda, skipBuild, noBundle, updateBinariesReleaseMetadataAndSaveGitBundle
 }: DownloadCommandArgs) {
+    const useBundle = noBundle != true;
     const octokit = new Octokit();
     const [githubOwner, githubRepo] = repo.split("/");
 
@@ -110,37 +114,45 @@ export async function DownloadLlamaCppCommand({
     type GithubReleaseType = Awaited<ReturnType<typeof octokit.rest.repos.getLatestRelease>> |
         Awaited<ReturnType<typeof octokit.rest.repos.getReleaseByTag>>;
 
-    let githubRelease: GithubReleaseType | null = null;
-    await withOra({
-        loading: chalk.blue("Fetching llama.cpp info"),
-        success: chalk.blue("Fetched llama.cpp info"),
-        fail: chalk.blue("Failed to fetch llama.cpp info")
-    }, async () => {
-        try {
-            if (release === "latest") {
-                githubRelease = await octokit.rest.repos.getLatestRelease({
-                    owner: githubOwner,
-                    repo: githubRepo
-                });
-            } else {
-                githubRelease = await octokit.rest.repos.getReleaseByTag({
-                    owner: githubOwner,
-                    repo: githubRepo,
-                    tag: release
-                });
+    let githubReleaseTag: string | null = (useBundle && (await getGitBundlePathForRelease(githubOwner, githubRepo, release)) != null)
+        ? release
+        : null;
+
+    if (githubReleaseTag == null)
+        await withOra({
+            loading: chalk.blue("Fetching llama.cpp info"),
+            success: chalk.blue("Fetched llama.cpp info"),
+            fail: chalk.blue("Failed to fetch llama.cpp info")
+        }, async () => {
+            let githubRelease: GithubReleaseType | null = null;
+
+            try {
+                if (release === "latest") {
+                    githubRelease = await octokit.rest.repos.getLatestRelease({
+                        owner: githubOwner,
+                        repo: githubRepo
+                    });
+                } else {
+                    githubRelease = await octokit.rest.repos.getReleaseByTag({
+                        owner: githubOwner,
+                        repo: githubRepo,
+                        tag: release
+                    });
+                }
+            } catch (err) {
+                console.error("Failed to fetch llama.cpp release info", err);
             }
-        } catch (err) {
-            console.error("Failed to fetch llama.cpp release info", err);
-        }
 
-        if (githubRelease == null) {
-            throw new Error(`Failed to find release "${release}" of "${repo}"`);
-        }
+            if (githubRelease == null) {
+                throw new Error(`Failed to find release "${release}" of "${repo}"`);
+            }
 
-        if (githubRelease!.data.tag_name == null) {
-            throw new Error(`Failed to find tag of release "${release}" of "${repo}"`);
-        }
-    });
+            if (githubRelease.data.tag_name == null) {
+                throw new Error(`Failed to find tag of release "${release}" of "${repo}"`);
+            }
+
+            githubReleaseTag = githubRelease.data.tag_name;
+        });
 
     await clearTempFolder();
 
@@ -153,7 +165,7 @@ export async function DownloadLlamaCppCommand({
     });
 
     console.log(chalk.blue("Cloning llama.cpp"));
-    await cloneLlamaCppRepo(githubOwner, githubRepo, githubRelease!.data.tag_name, noBundle != true);
+    await cloneLlamaCppRepo(githubOwner, githubRepo, githubReleaseTag!, useBundle);
 
     if (!skipBuild) {
         await downloadCmakeIfNeeded(true);
@@ -174,7 +186,7 @@ export async function DownloadLlamaCppCommand({
     }
 
     if (isCI && updateBinariesReleaseMetadataAndSaveGitBundle) {
-        await setBinariesGithubRelease(githubRelease!.data.tag_name);
+        await setBinariesGithubRelease(githubReleaseTag!);
         await unshallowAndSquashCurrentRepoAndSaveItAsReleaseBundle();
     }
 

--- a/src/config.ts
+++ b/src/config.ts
@@ -57,3 +57,8 @@ export const defaultChatSystemPrompt = "You are a helpful, respectful and honest
     "If you don't know the answer to a question, please don't share false information.";
 export const cliBinName = "node-llama-cpp";
 export const npxRunPrefix = "npx --no ";
+
+const documentationUrl = "https://withcatai.github.io/node-llama-cpp";
+export const documentationPageUrls = {
+    CUDA: documentationUrl + "/guide/CUDA"
+} as const;

--- a/src/utils/cloneLlamaCppRepo.ts
+++ b/src/utils/cloneLlamaCppRepo.ts
@@ -55,14 +55,35 @@ export async function cloneLlamaCppRepo(githubOwner: string, githubRepo: string,
         } catch (err) {
             await fs.remove(llamaCppDirectory);
             console.error("Failed to clone git bundle, cloning from GitHub instead", err);
+
+            printCloneErrorHelp(String(err));
         }
     }
 
-    await withGitCloneProgress("GitHub", async (gitWithCloneProgress) => {
-        await gitWithCloneProgress.clone(remoteGitUrl, llamaCppDirectory, {
-            "--depth": 1,
-            "--branch": tag,
-            "--quiet": null
+    try {
+        await withGitCloneProgress("GitHub", async (gitWithCloneProgress) => {
+            await gitWithCloneProgress.clone(remoteGitUrl, llamaCppDirectory, {
+                "--depth": 1,
+                "--branch": tag,
+                "--quiet": null
+            });
         });
-    });
+    } catch (err) {
+        printCloneErrorHelp(String(err));
+
+        throw err;
+    }
+}
+
+function printCloneErrorHelp(error: string) {
+    // This error happens with some docker images where the current user is different
+    // from the owner of the files due to mounting a volume.
+    // In such cases, print a helpful message to help the user resolve the issue.
+    if (error.toLowerCase().includes("detected dubious ownership in repository"))
+        console.info("\n" +
+            chalk.grey("[node-llama-cpp]") + chalk.yellow(" To fix this issue, try running this command to fix it for the current module directory:") + "\n" +
+            'git config --global --add safe.directory "' + llamaCppDirectory + '"\n\n' +
+            chalk.yellow("Or run this command to fix it everywhere:") + "\n" +
+            'git config --global --add safe.directory "*"'
+        );
 }

--- a/src/utils/cmake.ts
+++ b/src/utils/cmake.ts
@@ -35,6 +35,12 @@ export async function getCmakePath() {
 
         if (resolvedPath.toLowerCase().endsWith(".cmd"))
             resolvedPath = (await getBinFromWindowCmd(resolvedPath, "cmake.exe")) ?? "";
+        else if (resolvedPath.toLowerCase().endsWith(".ps1")) {
+            const cmdFilePath = resolvedPath.slice(0, -".ps1".length) + ".cmd";
+
+            if (await fs.pathExists(cmdFilePath))
+                resolvedPath = (await getBinFromWindowCmd(cmdFilePath, "cmake.exe")) ?? "";
+        }
 
         if (resolvedPath !== "")
             return resolvedPath;

--- a/src/utils/compileLLamaCpp.ts
+++ b/src/utils/compileLLamaCpp.ts
@@ -24,35 +24,33 @@ export async function compileLlamaCpp({
         const cmakePathArgs = await getCmakePathArgs();
         const toolchainFile = await getToolchainFileForArch(arch);
         const runtimeVersion = nodeTarget.startsWith("v") ? nodeTarget.slice("v".length) : nodeTarget;
-        const cmakeCustomOptions = [];
+        const cmakeCustomOptions = new Map<string, string>();
 
-        if ((metal && process.platform === "darwin") || process.env.LLAMA_METAL === "1") cmakeCustomOptions.push("LLAMA_METAL=1");
-        else cmakeCustomOptions.push("LLAMA_METAL=OFF");
+        if ((metal && process.platform === "darwin") || process.env.LLAMA_METAL === "1") cmakeCustomOptions.set("LLAMA_METAL", "1");
+        else cmakeCustomOptions.set("LLAMA_METAL", "OFF");
 
-        if (cuda || process.env.LLAMA_CUBLAS === "1") cmakeCustomOptions.push("LLAMA_CUBLAS=1");
-        if (cuda && process.env.CUDA_PATH != null && await fs.pathExists(process.env.CUDA_PATH))
-            cmakeCustomOptions.push("CMAKE_GENERATOR_TOOLSET=" + process.env.CUDA_PATH);
+        if (cuda || process.env.LLAMA_CUBLAS === "1") cmakeCustomOptions.set("LLAMA_CUBLAS", "1");
 
-        if (process.env.LLAMA_MPI === "1") cmakeCustomOptions.push("LLAMA_MPI=1");
-        if (process.env.LLAMA_OPENBLAS === "1") cmakeCustomOptions.push("LLAMA_OPENBLAS=1");
-        if (process.env.LLAMA_BLAS_VENDOR != null) cmakeCustomOptions.push("LLAMA_BLAS_VENDOR=" + process.env.LLAMA_BLAS_VENDOR);
-        if (process.env.LLAMA_CUDA_FORCE_DMMV != null) cmakeCustomOptions.push("LLAMA_CUDA_FORCE_DMMV=" + process.env.LLAMA_CUDA_FORCE_DMMV);
-        if (process.env.LLAMA_CUDA_DMMV_X != null) cmakeCustomOptions.push("LLAMA_CUDA_DMMV_X=" + process.env.LLAMA_CUDA_DMMV_X);
-        if (process.env.LLAMA_CUDA_MMV_Y != null) cmakeCustomOptions.push("LLAMA_CUDA_MMV_Y=" + process.env.LLAMA_CUDA_MMV_Y);
-        if (process.env.LLAMA_CUDA_F16 != null) cmakeCustomOptions.push("LLAMA_CUDA_F16=" + process.env.LLAMA_CUDA_F16);
-        if (process.env.LLAMA_CUDA_KQUANTS_ITER != null) cmakeCustomOptions.push("LLAMA_CUDA_KQUANTS_ITER=" + process.env.LLAMA_CUDA_KQUANTS_ITER);
-        if (process.env.LLAMA_CUDA_PEER_MAX_BATCH_SIZE != null) cmakeCustomOptions.push("LLAMA_CUDA_PEER_MAX_BATCH_SIZE=" + process.env.LLAMA_CUDA_PEER_MAX_BATCH_SIZE);
-        if (process.env.LLAMA_HIPBLAS === "1") cmakeCustomOptions.push("LLAMA_HIPBLAS=1");
-        if (process.env.LLAMA_CLBLAST === "1") cmakeCustomOptions.push("LLAMA_CLBLAST=1");
+        if (process.env.LLAMA_MPI === "1") cmakeCustomOptions.set("LLAMA_MPI", "1");
+        if (process.env.LLAMA_OPENBLAS === "1") cmakeCustomOptions.set("LLAMA_OPENBLAS", "1");
+        if (process.env.LLAMA_BLAS_VENDOR != null) cmakeCustomOptions.set("LLAMA_BLAS_VENDOR", process.env.LLAMA_BLAS_VENDOR);
+        if (process.env.LLAMA_CUDA_FORCE_DMMV != null) cmakeCustomOptions.set("LLAMA_CUDA_FORCE_DMMV", process.env.LLAMA_CUDA_FORCE_DMMV);
+        if (process.env.LLAMA_CUDA_DMMV_X != null) cmakeCustomOptions.set("LLAMA_CUDA_DMMV_X", process.env.LLAMA_CUDA_DMMV_X);
+        if (process.env.LLAMA_CUDA_MMV_Y != null) cmakeCustomOptions.set("LLAMA_CUDA_MMV_Y", process.env.LLAMA_CUDA_MMV_Y);
+        if (process.env.LLAMA_CUDA_F16 != null) cmakeCustomOptions.set("LLAMA_CUDA_F16", process.env.LLAMA_CUDA_F16);
+        if (process.env.LLAMA_CUDA_KQUANTS_ITER != null) cmakeCustomOptions.set("LLAMA_CUDA_KQUANTS_ITER", process.env.LLAMA_CUDA_KQUANTS_ITER);
+        if (process.env.LLAMA_CUDA_PEER_MAX_BATCH_SIZE != null) cmakeCustomOptions.set("LLAMA_CUDA_PEER_MAX_BATCH_SIZE", process.env.LLAMA_CUDA_PEER_MAX_BATCH_SIZE);
+        if (process.env.LLAMA_HIPBLAS === "1") cmakeCustomOptions.set("LLAMA_HIPBLAS", "1");
+        if (process.env.LLAMA_CLBLAST === "1") cmakeCustomOptions.set("LLAMA_CLBLAST", "1");
 
         if (toolchainFile != null)
-            cmakeCustomOptions.push("CMAKE_TOOLCHAIN_FILE=" + toolchainFile);
+            cmakeCustomOptions.set("CMAKE_TOOLCHAIN_FILE", toolchainFile);
 
         for (const key in process.env) {
             if (key.startsWith(customCmakeOptionsEnvVarPrefix)) {
                 const option = key.slice(customCmakeOptionsEnvVarPrefix.length);
                 const value = process.env[key];
-                cmakeCustomOptions.push(`${option}=${value}`);
+                cmakeCustomOptions.set(option, value!);
             }
         }
 
@@ -63,7 +61,7 @@ export async function compileLlamaCpp({
         await spawnCommand(
             "npm",
             ["run", "-s", "cmake-js-llama", "--", "compile", "--log-level", "warn", "--arch=" + arch, "--runtime-version=" + runtimeVersion, ...cmakePathArgs]
-                .concat(cmakeCustomOptions.map(option => "--CD" + option)),
+                .concat([...cmakeCustomOptions].map(([key, value]) => "--CD" + key + "=" + value)),
             __dirname
         );
 

--- a/src/utils/compileLLamaCpp.ts
+++ b/src/utils/compileLLamaCpp.ts
@@ -2,7 +2,10 @@ import path from "path";
 import {fileURLToPath} from "url";
 import process from "process";
 import fs from "fs-extra";
-import {customCmakeOptionsEnvVarPrefix, llamaCppDirectory, llamaDirectory, llamaToolchainsDirectory} from "../config.js";
+import chalk from "chalk";
+import {
+    customCmakeOptionsEnvVarPrefix, documentationPageUrls, llamaCppDirectory, llamaDirectory, llamaToolchainsDirectory
+} from "../config.js";
 import {clearLlamaBuild} from "./clearLlamaBuild.js";
 import {setUsedBinFlag} from "./usedBinFlag.js";
 import {spawnCommand} from "./spawnCommand.js";
@@ -86,6 +89,13 @@ export async function compileLlamaCpp({
     } catch (err) {
         if (setUsedBinFlagArg)
             await setUsedBinFlag("prebuiltBinaries");
+
+        if (cuda)
+            console.info("\n" +
+                chalk.grey("[node-llama-cpp] ") +
+                chalk.yellow("To resolve errors related to CUDA compilation, see the CUDA guide: ") +
+                documentationPageUrls.CUDA
+            );
 
         throw err;
     } finally {

--- a/src/utils/gbnfJson/terminals/GbnfStringValue.ts
+++ b/src/utils/gbnfJson/terminals/GbnfStringValue.ts
@@ -11,14 +11,16 @@ export class GbnfStringValue extends GbnfTerminal {
 
     override getGrammar(): string {
         return [
-            "\"",
+            '"',
+            '\\"',
             this.value
                 .replaceAll("\\", "\\\\")
                 .replaceAll("\t", "\\t")
                 .replaceAll("\r", "\\r")
                 .replaceAll("\n", "\\n")
                 .replaceAll('"', "\\\\" + '\\"'),
-            "\""
+            '\\"',
+            '"'
         ].join("");
     }
 }

--- a/test/standalone/llamaEvaluator/LlamaGrammar.test.ts
+++ b/test/standalone/llamaEvaluator/LlamaGrammar.test.ts
@@ -104,7 +104,7 @@ describe("grammar for JSON schema", () => {
         };
 
         expect(grammar.grammar).toMatchInlineSnapshot(`
-          "root ::= \\"{\\" whitespace-new-lines-rule \\"message\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"numberOfWordsInMessage\\" \\":\\" [ ]? integer-number-rule \\",\\" whitespace-new-lines-rule \\"feelingGoodPercentage\\" \\":\\" [ ]? fractional-number-rule \\",\\" whitespace-new-lines-rule \\"feelingGood\\" \\":\\" [ ]? boolean-rule \\",\\" whitespace-new-lines-rule \\"feelingOverall\\" \\":\\" [ ]? rule5 \\",\\" whitespace-new-lines-rule \\"verbsInMessage\\" \\":\\" [ ]? rule6 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
+          "root ::= \\"{\\" whitespace-new-lines-rule \\"\\\\\\"message\\\\\\"\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"\\\\\\"numberOfWordsInMessage\\\\\\"\\" \\":\\" [ ]? integer-number-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"feelingGoodPercentage\\\\\\"\\" \\":\\" [ ]? fractional-number-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"feelingGood\\\\\\"\\" \\":\\" [ ]? boolean-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"feelingOverall\\\\\\"\\" \\":\\" [ ]? rule5 \\",\\" whitespace-new-lines-rule \\"\\\\\\"verbsInMessage\\\\\\"\\" \\":\\" [ ]? rule6 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
           whitespace-new-lines-rule ::= [\\\\n]? [ \\\\t]* [\\\\n]?
           string-rule ::= \\"\\\\\\"\\" ( [^\\"\\\\\\\\] | \\"\\\\\\\\\\" ([\\"\\\\\\\\/bfnrt] | \\"u\\" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]))* \\"\\\\\\"\\"
           null-rule ::= \\"null\\"
@@ -114,8 +114,8 @@ describe("grammar for JSON schema", () => {
           rule1 ::= \\"true\\"
           rule2 ::= \\"false\\"
           boolean-rule ::= ( rule1 | rule2 )
-          rule3 ::= \\"good\\"
-          rule4 ::= \\"bad\\"
+          rule3 ::= \\"\\\\\\"good\\\\\\"\\"
+          rule4 ::= \\"\\\\\\"bad\\\\\\"\\"
           rule5 ::= ( rule3 | rule4 )
           rule7 ::= ( string-rule ) ( \\",\\" whitespace-new-lines-rule string-rule )*
           rule8 ::= ( string-rule )?
@@ -210,7 +210,7 @@ describe("grammar for JSON schema", () => {
           "root ::= \\"[\\" whitespace-new-lines-rule ( rule2 | rule3 ) whitespace-new-lines-rule \\"]\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
           whitespace-new-lines-rule ::= [\\\\n]? [ \\\\t]* [\\\\n]?
           string-rule ::= \\"\\\\\\"\\" ( [^\\"\\\\\\\\] | \\"\\\\\\\\\\" ([\\"\\\\\\\\/bfnrt] | \\"u\\" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]))* \\"\\\\\\"\\"
-          rule0 ::= \\"{\\" whitespace-new-lines-rule \\"message\\" \\":\\" [ ]? string-rule whitespace-new-lines-rule \\"}\\"
+          rule0 ::= \\"{\\" whitespace-new-lines-rule \\"\\\\\\"message\\\\\\"\\" \\":\\" [ ]? string-rule whitespace-new-lines-rule \\"}\\"
           rule1 ::= ( rule0 | string-rule )
           rule2 ::= ( rule1 ) ( \\",\\" whitespace-new-lines-rule rule1 )*
           rule3 ::= ( rule1 )?"
@@ -297,12 +297,12 @@ describe("grammar for JSON schema", () => {
         };
 
         expect(grammar.grammar).toMatchInlineSnapshot(`
-          "root ::= \\"{\\" whitespace-new-lines-rule \\"onlyPositiveText\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"onlyNegativeText\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"onlyVibe\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"onlyNumber\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"worstThing\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"withNewLine\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"withQuotes\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
+          "root ::= \\"{\\" whitespace-new-lines-rule \\"\\\\\\"onlyPositiveText\\\\\\"\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNegativeText\\\\\\"\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyVibe\\\\\\"\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNumber\\\\\\"\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"worstThing\\\\\\"\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"withNewLine\\\\\\"\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"\\\\\\"withQuotes\\\\\\"\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
           whitespace-new-lines-rule ::= [\\\\n]? [ \\\\t]* [\\\\n]?
-          rule0 ::= \\"good\\"
+          rule0 ::= \\"\\\\\\"good\\\\\\"\\"
           null-rule ::= \\"null\\"
-          rule1 ::= \\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\"
-          rule2 ::= \\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\""
+          rule1 ::= \\"\\\\\\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\\\\\"\\"
+          rule2 ::= \\"\\\\\\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\\\\\"\\""
         `);
 
         const parsedValue = grammar.parse(JSON.stringify(exampleValidValue));
@@ -374,12 +374,12 @@ describe("grammar for JSON schema", () => {
         };
 
         expect(grammar.grammar).toMatchInlineSnapshot(`
-          "root ::= \\"{\\" whitespace-new-lines-rule \\"onlyPositiveText\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"onlyNegativeText\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"onlyVibe\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"onlyNumber\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"worstThing\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"withNewLine\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"withQuotes\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
+          "root ::= \\"{\\" whitespace-new-lines-rule \\"\\\\\\"onlyPositiveText\\\\\\"\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNegativeText\\\\\\"\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyVibe\\\\\\"\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNumber\\\\\\"\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"worstThing\\\\\\"\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"withNewLine\\\\\\"\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"\\\\\\"withQuotes\\\\\\"\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
           whitespace-new-lines-rule ::= [\\\\n]? [ \\\\t]* [\\\\n]?
-          rule0 ::= \\"good\\"
+          rule0 ::= \\"\\\\\\"good\\\\\\"\\"
           null-rule ::= \\"null\\"
-          rule1 ::= \\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\"
-          rule2 ::= \\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\""
+          rule1 ::= \\"\\\\\\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\\\\\"\\"
+          rule2 ::= \\"\\\\\\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\\\\\"\\""
         `);
 
         const parsedValue = grammar.parse(JSON.stringify(exampleValidValue));
@@ -452,12 +452,12 @@ describe("grammar for JSON schema", () => {
         };
 
         expect(grammar.grammar).toMatchInlineSnapshot(`
-          "root ::= \\"{\\" whitespace-new-lines-rule \\"onlyPositiveText\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"onlyNegativeText\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"onlyVibe\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"onlyNumber\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"worstThing\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"withNewLine\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"withQuotes\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
+          "root ::= \\"{\\" whitespace-new-lines-rule \\"\\\\\\"onlyPositiveText\\\\\\"\\" \\":\\" [ ]? \\"true\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNegativeText\\\\\\"\\" \\":\\" [ ]? \\"false\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyVibe\\\\\\"\\" \\":\\" [ ]? rule0 \\",\\" whitespace-new-lines-rule \\"\\\\\\"onlyNumber\\\\\\"\\" \\":\\" [ ]? \\"10\\" \\",\\" whitespace-new-lines-rule \\"\\\\\\"worstThing\\\\\\"\\" \\":\\" [ ]? null-rule \\",\\" whitespace-new-lines-rule \\"\\\\\\"withNewLine\\\\\\"\\" \\":\\" [ ]? rule1 \\",\\" whitespace-new-lines-rule \\"\\\\\\"withQuotes\\\\\\"\\" \\":\\" [ ]? rule2 whitespace-new-lines-rule \\"}\\" [\\\\n] [\\\\n] [\\\\n] [\\\\n] [\\\\n]*
           whitespace-new-lines-rule ::= [\\\\n]? [ \\\\t]* [\\\\n]?
-          rule0 ::= \\"good\\"
+          rule0 ::= \\"\\\\\\"good\\\\\\"\\"
           null-rule ::= \\"null\\"
-          rule1 ::= \\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\"
-          rule2 ::= \\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\""
+          rule1 ::= \\"\\\\\\"Hooray!\\\\nYes!\\\\t/\\\\\\\\\\\\\\"\\"
+          rule2 ::= \\"\\\\\\"The message is \\\\\\\\\\\\\\"Hi!\\\\\\\\\\\\\\".\\\\\\"\\""
         `);
 
         const parsedValue = grammar.parse(JSON.stringify(exampleValidValue));

--- a/vitest.config.ts
+++ b/vitest.config.ts
@@ -1,0 +1,7 @@
+import {defineConfig} from "vitest/config";
+
+export default defineConfig({
+    test: {
+        threads: false
+    }
+});

